tket/src/Architecture/Architecture.hpp

```diff
@@ -48,44 +49,77 @@ static std::vector<std::pair<Node, Node>> as_nodepairs(
   return nodepairs;
 }
 
-class Architecture : public graphs::UIDConnectivity<Node> {
- private:
-  using Base = graphs::UIDConnectivity<Node>;
+/**
+ * Generic architecture class
+ *
+ * Constraint: the underlying graph type must have node type @ref Node.
+ *
+ * @tparam T underlying graph type
+ */
+template <typename T>
+class ArchitectureBase : public T {
+ public:
+  using T::edge_exists;
+  using T::get_all_edges_vec;
+  using T::get_all_nodes_vec;
+  using T::n_nodes;
+  using T::node_exists;
+  using T::nodes;
+  using T::T;
+};
 
+class Architecture : public ArchitectureBase<graphs::DirectedGraph<Node>> {
  public:
-  /** inherit constructors */
-  using Base::Base;
+  /** Construct an empty Architecture. */
+  Architecture() : ArchitectureBase() {}
+
+  /** Construct an Architecture with given nodes and no edges. */
+  explicit Architecture(const std::vector<Node> &nodes)
+      : ArchitectureBase(nodes) {}
+
+  /** Construct an Architecture with given edges. */
+  explicit Architecture(const std::vector<std::pair<Node, Node>> &edges)
+      : ArchitectureBase(edges) {}
 
   /**
    * Construct from a vector of pairs of indices in the default register.
    */
   explicit Architecture(const std::vector<std::pair<unsigned, unsigned>> &edges)
       : Architecture(as_nodepairs(edges)) {}
 
-  /** compute articulation points of Architecture
-   *  if `subarc` is given, the articulation that do not affect the
-   * sub-architecture are ignored */
+  /**
+   * Nodes that cannot be removed without breaking connectivity.
+   */
   node_set_t get_articulation_points() const;
+
+  /**
+   * Nodes that cannot be removed without breaking connectivity of the subarc.
+   */
   node_set_t get_articulation_points(const Architecture &subarc) const;
 
-  /* returns new Architecture that is generated by a subset of nodes of `this`
+  /**
+   * Sub-architecture generated by a subset of nodes.
    */
   Architecture create_subarch(const std::vector<Node> &nodes);
 
-  // Returns vectors of nodes which correspond to lines of specified length
+  /**
+   * Vectors of nodes corresponding to lines of given lengths
+   */
   std::vector<node_vector_t> get_lines(
       std::vector<unsigned> required_lengths) const;
 
-  // get architecture diameter
+  /**
+   * Diameter of graph
+   */
   unsigned get_diameter() const;
```

Comment:
It might make sense to move some of these methods (at least the basic ones such as `get_diameter` and `create_subarch`) to be part of a common API in `ArchitectureBase`.

Reply:
I have moved `get_diameter` into `AbstractGraph`. I don't think `create_subarch` should be moved down because a subset of nodes of a given architecture type won't necessarily be of that type (e.g. a `RingArch`).